Keep local hour across daylight saving time change

unka65 opened the pull request:
For recurring type, the next schedule time does not keep the same local time hour if the current time offset is a different offset than the initial setup time due to daylight time saving change. This change uses local time to determine the next schedule day.

The schedule entry is updated with the next date/time to always use the current time change policy, to eliminate overhead of calculating it each reference and so can view in the schedule entry. The initial date/time is not retained. For tracking, the schedule date/time update is logged.

To verify the calculation, set a schedule with the date at least 2 days before the previous daylight saving time change. Refresh the list. The schedule will have the next schedule date with the same local hour. The logic does not work for a schedule for only the hour of the time change where the offset increases (daylight saving time to standard time).

Additional changes not included to consider are for the columns in the list to show more detail of the next interval.

Existing
![image](https://github.com/Cacti/plugin_maint/assets/61275958/9c815c13-180b-4649-8629-6e1943d4f816)

Suggested.  MM/DD is fixed format, not based on locale.  Change Interval from Not Defined.
![image](https://github.com/Cacti/plugin_maint/assets/61275958/8a928df7-64fb-464c-9c00-b2cc19b18004)

unka65 committed:
Keep local hour across daylight saving time change

For recurring type, the next schedule time does not keep the same local time hour if the current time offset is a different offset than the initial setup time due to daylight time saving change. This change uses local time to determine the next schedule day.

The schedule entry is updated with the next date/time to always use the current time change policy, to eliminate overhead of calculating it each reference and so can view in the schedule entry. The initial date/time is not retained. For tracking, the schedule date/time update is logged.

To verify the calculation, set a schedule with the date at least 2 days before the previous daylight saving time change. Refresh the list. The schedule will have the next schedule date with the same local hour. The logic does not work for a schedule for only the hour of the time change where the offset increases (daylight saving time to standard time).

xmacan commented on functions.php:
```diff
@@ -61,13 +61,35 @@ function plugin_maint_check_schedule($schedule) {
+				/* past, calculate next */
+				if ($sc['etime'] < $t) {
+					/* convert start and end to local so that hour stays same for add days across daylight saving time change */
+					$starttimelocal = (new DateTimeImmutable('@' . strval($sc['stime'])))->setTimezone( new DateTimeZone( date_default_timezone_get()));
```
DateTimeImmutable class is for PHP 5 >= 5.5.0. Cacti 1.2.x supports PHP 5.4+. So there is small version inconsistency.  Anyone using old php like this will probably have problems elsewhere

TheWitness replied:
@unka65, Can you please make a modification to perform a `version_compare()` here?

unka65 replied:
I did not realize older PHP was still supported.  DateTimeImmutable is not needed and so is removed.  Let me know if that does not address it.

unka65 committed:
Use DateTime

Newer DateTimeImmutable is not needed.  I did not realize older PHP was still supported.

TheWitness commented:
Appreciate your contribution!
